ulib/FStar.LexicographicOrdering.fsti

```diff
@@ -0,0 +1,135 @@
+(*
+   Copyright 2021 Microsoft Research
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+   Authors: Aseem Rastogi and Nikhil Swamy
+*)
+
+module FStar.LexicographicOrdering
+
+/// This module proves that lexicographic and symmetric products are
+///   well-founded (i.e. every element is accessible)
+///
+/// The main theorems in the module are `lex_wf` and `sym_wf`
+///
+/// See tests/micro-benchmarks/Test.WellFoundedRecursion.fst for how we use `lex` to prove termination for the ackermann function
+/// 
+/// Some references:
+///   - https://github.com/coq/coq/blob/master/theories/Wellfounded/Lexicographic_Product.v
+///   - Constructing Recursion Operators in Type Theory, L. Paulson  JSC (1986) 2, 325-355
+
+open FStar.Preorder
+open FStar.WellFounded
+
+
+/// Definition of lexicographic ordering as a relation over dependent tuples
+///
+/// Two elements are related if:
+///   - Either their first components are related
+///   - Or, the first components are equal, and the second components are related
+
+noeq
+type lex_t (#a:Type) (#b:a -> Type) (r_a:relation a) (r_b:(x:a -> relation (b x)))
+  : (x:a & b x) -> (x:a & b x) -> Type =
+  | Left_lex:
+    x1:a -> x2:a ->
+    y1:b x1 -> y2:b x2 ->
+    r_a x1 x2 ->
+    lex_t r_a r_b (| x1, y1 |) (| x2, y2 |)
+  | Right_lex:
+    x:a ->
+    y1:b x -> y2:b x ->
+    r_b x y1 y2 ->
+    lex_t r_a r_b (| x, y1 |) (| x, y2 |)
+
+/// Given two well-founded relations `r_a` and `r_b`,
+///   their lexicographic ordering is also well-founded
+
+val lex_t_wf (#a:Type) (#b:a -> Type)
+  (#r_a:relation a)
+  (#r_b:(x:a -> relation (b x)))
+  (wf_a:well_founded r_a)
+  (wf_b:(x:a -> well_founded (r_b x)))
+  : well_founded (lex_t r_a r_b)
+
+
+/// We can also define a squashed version of lex relation
+
+unfold
+let lex_aux (#a:Type) (#b:a -> Type)
+  (r_a:relation a)
+  (r_b:(x:a -> relation (b x)))
+  : relation (x:a & b x)
+  = fun (| x1, y1 |) (| x2, y2 |) ->
+    (squash (r_a x1 x2)) \/
+    (x1 == x2 /\ squash ((r_b x1) y1 y2))
+
+
+/// Provide a mapping from a point in lex_sq to a squashed point in lex
```

Comment:
nit: I guess you mean `lex_aux` rather than `lex_sq`

Reply:
Thanks, fixed!